hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/balancer/Dispatcher.java

```diff
@@ -848,16 +867,30 @@ private long getBlockList() throws IOException {
           synchronized (block) {
             block.clearLocations();
 
+            if (blkLocs instanceof StripedBlockWithLocations) {
+              // indices may have been adjusted before
+              ((DBlockStriped) block).setIndices(
+                  ((StripedBlockWithLocations) blkLocs).getIndices());
+            }
+
             // update locations
+            List<Integer> adjustList = new ArrayList<>();
             final String[] datanodeUuids = blkLocs.getDatanodeUuids();
             final StorageType[] storageTypes = blkLocs.getStorageTypes();
             for (int i = 0; i < datanodeUuids.length; i++) {
               final StorageGroup g = storageGroupMap.get(
                   datanodeUuids[i], storageTypes[i]);
               if (g != null) { // not unknown
                 block.addLocation(g);
+              } else if (blkLocs instanceof StripedBlockWithLocations) {
+                adjustList.add(i);
               }
             }
+
+            if (blkLocs instanceof StripedBlockWithLocations) {
+              // adjust indices if locations has been updated
```

Comment:
Could you please provide more detailed comments on when the locations could be updated and why we need to adjust indices?

Reply:
This is because `waitForBalancer` not waiting for namenode to delete extra replications. 
I fix the code and check the total block counts before `verifyLocatedStripedBlocks`

Reply:
I fix the code and add more comments like this.
```java
    if (!adjustList.isEmpty()) {
      // block.locations mismatch with block.indices
      // adjust indices to get correct internalBlock for Datanode in #getInternalBlock
      ((DBlockStriped) block).adjustIndices(adjustList);
      Preconditions.checkArgument(((DBlockStriped) block).indices.length
          == block.locations.size());
    }
```